flink-connectors/flink-connector-elasticsearch7/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchWriter.java

```diff
@@ -0,0 +1,326 @@
+        }
+
+        BackoffPolicy backoffPolicy;
+        if (bulkProcessorConfig.getFlushBackoffType() != null) {
+            final TimeValue backoffDelay =
+                    new TimeValue(bulkProcessorConfig.getBulkFlushBackOffDelay());
+            final int maxRetryCount = bulkProcessorConfig.getBulkFlushBackoffRetries();
+            switch (bulkProcessorConfig.getFlushBackoffType()) {
+                case CONSTANT:
+                    backoffPolicy = BackoffPolicy.constantBackoff(backoffDelay, maxRetryCount);
+                    break;
+                case EXPONENTIAL:
+                default:
+                    backoffPolicy = BackoffPolicy.exponentialBackoff(backoffDelay, maxRetryCount);
+            }
+        } else {
+            backoffPolicy = BackoffPolicy.noBackoff();
+        }
+        builder.setBackoffPolicy(backoffPolicy);
+
+        return builder.build();
+    }
+
+    private class BulkListener implements BulkProcessor.Listener {
+
+        @Override
+        public void beforeBulk(long executionId, BulkRequest request) {
+            LOG.info("Sending bulk of {} actions to Elasticsearch.", request.numberOfActions());
+            lastSendTime = System.currentTimeMillis();
+            numBytesOutCounter.inc(request.estimatedSizeInBytes());
+        }
+
+        @Override
+        public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+            receiveTime = System.currentTimeMillis();
+            // If the writer is cancelled before the last bulk response (i.e. no flush on checkpoint
+            // configured or shutdown without a final
+            // checkpoint) the mailbox might already be shutdown, so we should not enqueue any
+            // actions.
+            if (isClosed()) {
+                return;
+            }
+            mailboxExecutor.execute(
+                    () -> {
+                        if (!response.hasFailures() && flushOnCheckpoint) {
+                            pendingActions -= request.numberOfActions();
+                            return;
+                        }
+
+                        for (int i = 0; i < response.getItems().length; i++) {
+                            final BulkItemResponse itemResponse = response.getItems()[i];
+                            if (!itemResponse.isFailed()) {
+                                continue;
+                            }
+                            final Throwable failure = itemResponse.getFailure().getCause();
+                            if (failure == null) {
+                                continue;
+                            }
+                            final RestStatus restStatus = itemResponse.getFailure().getStatus();
+                            final DocWriteRequest<?> actionRequest = request.requests().get(i);
+
+                            if (restStatus == null) {
+                                throw new FlinkRuntimeException(
```

Comment:
Maybe, we should collect all issues and use `ExceptionUtils#firstOrSuppressed`?

Reply:
Good idea 👍